website/blog/2023-4-10-incrementals-pt1.md

```diff
@@ -0,0 +1,140 @@
+Corso implementation, so we thought we’d share them here.
+
+<!-- truncate -->
+
+## Background
+
+Before we dive into the details of how incremental backups work, it’s useful to
+have some knowledge of how delta queries work in the Microsoft Graph API and how
+data is laid out in Corso backups.
+
+### Microsoft delta queries
+
+Microsoft provides a delta query API that allows developers to get only the
+changes to the endpoint since the last query was made. The API represents the
+idea of the “last query” with an opaque token that's returned when the set of
+items is done being listed. For example, if a developer wants to get a delta
+token for a specific email folder, the developer would first list all the items
+in the folder using the delta endpoint. On the final page of item results from
+the endpoint, the Graph API would return a token that could be used to retrieve
+future updates.
+
+All returned tokens represent a point in time and are independent from each
+other. This means that getting token a1 at time t1, making some changes, and
+then getting another token a2 at time t2 would give distinct tokens. Requesting
+the changes from token a1 would always give the changes made after time t1
+including those after time t2. Requesting changes from token a2 would give only
+the changes made after time t2. Tokens eventually expire though, so waiting a
+long time between backups (e.x. weeks) may cause all items to be enumerated
+again.
+
+## Corso full backups, incremental backups, and backup layout
+
+Before we get into the nuts and bolts of how Corso uses the Microsoft delta
+query API, it’s important to first define what’s in a backup and the terminology
+we’ll be using throughout this post.
+
+### Backup layout
+
+Internally, a single Corso backup consists of three main parts: a kopia manifest
+that Corso uses as the root object of the backup (BackupModel), a kopia snapshot
+of indexing information for Corso, and a kopia snapshot of the item data in the
+backup. The BackupModel contains summary information about the status of the
+backup (did it have errors, how many items did it backup, etc) and pointers to
+the two snapshots that contain information.
+
+The snapshot with indexing information contains the data output during a
+`corso backup details` command and is used to filter the set of restored items
+during restore commands. The snapshot contains one entry for every backed up
+M365 item in the backup.
+
+The snapshot of item data contains the raw bytes that Corso backed up from M365.
+Internally, Corso uses a file hierarchy in kopia that closely mirrors the layout
+of the data in M365. For example, if the user has a file in the OneDrive folder
+`work/important` then Corso creates a kopia path
+`<tenant ID>/onedrive/<user ID>/files/<drive ID>/root/work/important` for that
+file.
+
+Corso also stores a few extra bits of metadata to help with incremental backups.
```

Comment:
Where does it store this? Part of the index snap or data snap?

Reply:
they're stored in the data snapshot

Reply:
updated with this info